actionpack/lib/action_dispatch/journey/visitors.rb

```diff
@@ -167,32 +167,67 @@ def visit(node, block)
         INSTANCE = new
       end
 
-      class String < FunctionalVisitor # :nodoc:
-        private
-          def binary(node, seed)
+      class String # :nodoc:
+        def accept(node, seed)
+          case node.type
+          when :DOT
+            seed << node.left
+          when :LITERAL
+            seed << node.left
+          when :SYMBOL
+            seed << node.left
+          when :SLASH
+            seed << node.left
+          when :CAT
             visit(node.right, visit(node.left, seed))
-          end
-
-          def nary(node, seed)
+          when :STAR
+            visit(node.left, seed)
+          when :OR
             last_child = node.children.last
-            node.children.inject(seed) { |s, c|
-              string = visit(c, s)
-              string << "|" unless last_child == c
-              string
-            }
-          end
-
-          def terminal(node, seed)
-            seed + node.left
+            node.children.each do |c|
+              visit(c, seed)
+              seed << "|" unless last_child == c
+            end
+            seed
+          when :GROUP
+            visit(node.left, seed << "(") << ")"
+          else
+            raise "Unknown node type: #{node.type}"
           end
+        end
 
-          def visit_GROUP(node, seed)
-            visit(node.left, seed.dup << "(") << ")"
-          end
+        INSTANCE = new
 
-          INSTANCE = new
+        private
+          alias_method :visit, :accept
```

Comment:
Definitely :+1: to not re-allocating with `+`, I saw that as well.

I was looking into double dispatch visiting where nodes define

```
def visit(v, seed)
  v.visit_group(seed)
end
```

but I think yours would be faster because there's less indirection? I'll try out a benchmark to verify.

Anyways, is it worth even having this alias? We could just always call `accept` inside the visitor

Reply:
Yes, at this point the methods aren't doing much, so the bottleneck is the method call itself.

Hence why I went with a case with inline snippets that reduce the number of method calls.

I also have a non-recursive version, that push and pop the nodes on an array, but apparently it's slower.